keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,882 @@
+    CIDR     string
+
+    # This defines which nodes the config is applicable to. An empty selector
+    # can be applied to any node.
+    # +optional
+    NodeSelector    *v1.LabelSelector
+
+    # Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    # Users would have to ensure that the kubelet doesn't try to schedule
+    # more pods than are supported by the node's netmask (i.e. the kubelet's
+    # --max-pods flag)
+    # +required
+    PerNodeMaskSize int
+}
+```
+
+#### Expected Behavior
+
+-   Each node will be assigned up to one range from each `FamilyType`.
+    -   **TODO:** what is the tie-break if multiple ranges match
+-   An empty `NodeSelector` functions as a default that applies to all nodes.
+    This should be the fall-back and not take precedence if any other range
+    matches.
+-   `CIDR`, and `PerNodeMaskSize` are immutable after creation.
+-   Reject any changes that would change existing node allocations:
+    -   Can not delete a `CIDRRangeConfig` that is in use by a node
+    -   Can not modify a `NodeSelector` to deselect a node that is using that
+        range.
+
+#### Example Config
+
+```
+[
+{
+    # For existing clusters this is the same as ClusterCIDR
+    CIDR:            "10.0.0.0/8",
+    # Default for nodes not matching any other rule
+    NodeSelector:    nil,
+    # For existing API this is the same as NodeCIDRMaskSize
+    PerNodeMaskSize: 24,
+},
+{
+    CIDR:            "172.16.0.0/14",
+    # Another range, also allocate-able to any node
+    NodeSelector:    nil,
+    PerNodeMaskSize: 24,
+},
+{
+    CIDR:            "10.0.0.0/8",
+    NodeSelector:    { key: "np" op: "IN" value:["np1"] },
+    PerNodeMaskSize: 26,
+},
+{
+    CIDR:            "192.168.0.0/16",
+    NodeSelector:    { key: "np" op: "IN" value:["np2"] },
+    PerNodeMaskSize: 26,
+},
+{
+    CIDR:            "5.2.0.0/16",
+    NodeSelector:    { "np": "np3" },
+    PerNodeMaskSize: 20,
+},
+{
+    CIDR:            "fd12:3456:789a:1::/64",
+    NodeSelector:    { "np": "np3" },
+    PerNodeMaskSize: 112,
+},
+...
+]
+```
+
+Given the above config, a valid potential configuration might be:
+
+```
+{“np”: “np1”} --> "10.0.0.0/26"
+{“np”: “np2”} --> "192.16.0.0/26"
+{“np”: “np3”} --> "5.2.0.0/20", "fd12:3456:789a:1::/112"
+{“np”: “np4”} --> "172.16.0.0/24"
+```
+
+### Controller
+
+Implement a new
+[NodeIPAM controller](https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/nodeipam)
+The controller will set up watchers on the `ClusterCIDRConfig` objects and the
+`Node` objects.
+
+#### Data Structures
+
+We will use maps to store the allocated ranges and which node is using the
+range. Because the number of nodes is expected to be on the order of thousands,
+more sophisticated data structure are likely not required.
+
+Prior investgations [here](https://github.com/kubernetes/kubernetes/pull/90184)
+suggest that maps storing allocations will perform well under the number of
+nodes we expect.
+
+#### Startup
```

Comment:
kubeadm follows a "minimal viable cluster" model, that often suits both a basic and an advanced use case.
but the need to switch a cluster to a bigger CIDR can be a problem for all users.

ideally the proposal in this KEP, should ensure backwards compatibility with the existing configuration surface (which today is passing flags to core components) to configure a single CIDR. if some flags will be deprecated / removed, we should follow the official policy and alert users. migration guides should be written, et cetera.

kubeadm will likely continue to deploy one single arbitrary size CIDR by default.

Reply:
For now we plan to keep the flags. I've started outlining this in the Upgrade/downgrade stories, but users would just change their controller type and keep using all the same other configurations.

Maybe down the line (after this new controller and API has gone GA), we can revisit those flags, but I doubt deprecating them is high priority. As @neolit123 points out, we'll probably treat those flags as good starting points for users who "just want a cluster" and eventually the defaults will be setup so that they can seamlessly expand their CIDRs as their needs change.